Frankenflow: Customize checkout step

razvanpapadopol opened the pull request:
#### Changes proposed in this Pull Request

* hide Masterbar (blue top bar) from checkout page when `preLaunch` query param exists
* show Back button which redirects to plans step
* show custom header content

<img width="1260" alt="Screenshot 2020-03-02 at 18 19 48" src="https://user-images.githubusercontent.com/14192054/75700591-9d3a5000-5cb2-11ea-98a2-01ba453ad21a.png">

#### Testing instructions
* Go to `/start/frankenflow?siteSlug=SITE_SLUG` for a non-launched site and pick a plan
* On checkout page you should see a _Back_ button instead of the blue header and a header message saying your site is up and running and you should complete your plan purchase to upgrade.
* Pressing the _Back_ button should redirect to _Plans_ step

Fixes #38936

Commit:
Customize checkout step in Frankenflow.
- hide Masterbar from checkout page when preLaunch query param exists
- show Back button which redirects to plans step

Commit:
Add checkout header message and style back button.

razvanpapadopol commented on client/my-sites/checkout/checkout/checkout-container.jsx:
```diff
@@ -80,6 +84,16 @@ class CheckoutContainer extends React.Component {
+					>
+						<Gridicon icon="arrow-left" size={ 18 } />
+						{ this.props.translate( 'Back' ) }
+					</Button>
```
@simison Do you think we need to make this more generic at the moment?

simison replied:
I think it's fine as-is otherwise but I'd call things `isComingFromGutenboarding` since that's more clear to outsiders.

Commit:
Rename isComingFromFrankenflow prop to isComingFromGutenboarding

simison commented:
Let's tune up the header a bit, there's quite a lot of repetition there now:

<img width="1245" alt="Screenshot 2020-02-25 at 08 44 52" src="https://user-images.githubusercontent.com/87168/75221708-26c9b980-57ab-11ea-87f0-ee43fa0ecb30.png">

It's still much but better — we can then revisit when we review the copy for the whole flow.

sirreal reviewed (dismissed):
After purchasing a plan, there's no masterbar at `/home/SITE_SLUG`:

![Screen Shot 2020-02-25 at 08 47 50](https://user-images.githubusercontent.com/841763/75225849-a788b580-57ab-11ea-8a62-dbd1321db2b8.png)

yansern commented:
Pressing the Back button on the header goes back to the payment plans page, but pressing the browser back button shows me a blank grey page. The url of that grey page is `/start/frankenflow/launch`.

razvanpapadopol commented:
> After purchasing a plan, there's no masterbar at `/home/SITE_SLUG`:

Good catch, @sirreal ! Thanks for going through the whole flow 💯 

It seems the `showMasterbar()` call isn't happening [here](https://github.com/Automattic/wp-calypso/pull/39447/files#diff-7f164a3b7deeca0666c8602e2b20e4acR52) because of the redirect to home. So we'll follow the existing pattern in `client/layout/index` and set `isCheckoutFromGutenboarding` based on the checkout `sectionName` and query param.

Commit:
Replace masterbarToggle in checkout controller.
- update noMasterbarForRoute in client/layout
- add isCheckoutFromGutenboarding flag to identify the gutenboarding flow

razvanpapadopol commented:
> After purchasing a plan, there's no masterbar at `/home/SITE_SLUG`:

Good catch, @sirreal ! Thanks for going through the whole flow 💯 

It seems the `showMasterbar()` call isn't happening [here](https://github.com/Automattic/wp-calypso/pull/39447/files#diff-7f164a3b7deeca0666c8602e2b20e4acR52) because of the redirect to home. So we'll follow the existing pattern in `client/layout/index` and set `isCheckoutFromGutenboarding` based on the checkout `sectionName` and query param.

razvanpapadopol commented:
> After purchasing a plan, there's no masterbar at `/home/SITE_SLUG`:

Good catch, @sirreal ! Thanks for going through the whole flow 💯 

It seems the `showMasterbar()` call isn't happening [here](https://github.com/Automattic/wp-calypso/pull/39447/files#diff-7f164a3b7deeca0666c8602e2b20e4acR52) because of the redirect to home. So we'll follow the existing pattern in `client/layout/index` and set `isCheckoutFromGutenboarding` based on the checkout `sectionName` and query param.

Commit:
Tune up the checkout header when coming from Gutenboarding.
 - Avoid text repetition around site creation.
 - Add is-frankenflow class to section for further customization.

razvanpapadopol commented on client/my-sites/checkout/checkout/checkout-container.jsx:
```diff
@@ -73,15 +76,47 @@ class CheckoutContainer extends React.Component {
+									comment: '`siteSlug` is the WordPress.com site, e.g., testsite.wordpress.com',
+								} ) }
+							</div>
+						</div>
```
Not reusing much except the confetti image since all these will change as @simison stated in https://github.com/Automattic/wp-calypso/pull/39447#issuecomment-590710667

razvanpapadopol commented:
> Pressing the Back button on the header goes back to the payment plans page, but pressing the browser back button shows me a blank grey page. The url of that grey page is /start/frankenflow/launch.

@yansern this is the current behaviour also if going through the regular launch flow (using the _Launch_ button on home). While it's a valid point for Calypso standard flows, I wouldn't try to fix that in Frankenflow.

sirreal reviewed (commented):
This works well in my testing. A few notes about how I think this could be improved for clarity. It looks like there are a few places where we've got the same condition spread across different parts of the component and we could unify them.

Header looks a bit different from the description:

![Screen Shot 2020-02-27 at 16 59 28](https://user-images.githubusercontent.com/841763/75461522-37348c80-5983-11ea-9fbc-fc16040e9f41.png)

Commit:
use optional chaining

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

Commit:
Unify condition checks to improve clarity.
- use renderCheckoutHeader to render custom header based on isComingFromGutenboarding prop
- shouldDisplaySiteCreatedNotice now takes into consideration isComingFromGutenboarding prop

razvanpapadopol commented:
> It looks like there are a few places where we've got the same condition spread across different parts of the component and we could unify them.

Thanks for the suggestions! I've updated the existing methods and it looks much better indeed.

> Header looks a bit different from the description

Yes, it was fine tuned based on https://github.com/Automattic/wp-calypso/pull/39447#issuecomment-590710667

sirreal reviewed (approved):
Working great, thanks!
